pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go

```diff
@@ -20,47 +20,133 @@ import (
+func (pl *InterPodAffinity) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	var meta *predicates.PodAffinityMetadata
+	var allNodes []*nodeinfo.NodeInfo
+	var havePodsWithAffinityNodes []*nodeinfo.NodeInfo
+	var err error
+	if allNodes, err = pl.snapshotSharedLister.NodeInfos().List(); err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("failed to list NodeInfos: %v", err))
+	}
+	if havePodsWithAffinityNodes, err = pl.snapshotSharedLister.NodeInfos().HavePodsWithAffinityList(); err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("failed to list NodeInfos with pods with affinity: %v", err))
+	}
+	if meta, err = predicates.GetPodAffinityMetadata(pod, allNodes, havePodsWithAffinityNodes); err != nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("Error calculating podAffinityMetadata: %v", err))
+	}
+
+	s := &preFilterState{
+		meta: meta,
+	}
+	cycleState.Write(preFilterStateKey, s)
+	return nil
+}
+
+// PreFilterExtensions returns prefilter extensions, pod add and remove.
+func (pl *InterPodAffinity) PreFilterExtensions() framework.PreFilterExtensions {
+	return pl
+}
+
+// AddPod from pre-computed data in cycleState.
+func (pl *InterPodAffinity) AddPod(ctx context.Context, cycleState *framework.CycleState, podToSchedule *v1.Pod, podToAdd *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
+	meta, err := getPodAffinityMetadata(cycleState)
+	if err != nil || meta == nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	meta.UpdateWithPod(podToAdd, podToSchedule, nodeInfo.Node(), 1)
+	return nil
+}
+
+// RemovePod from pre-computed data in cycleState.
+func (pl *InterPodAffinity) RemovePod(ctx context.Context, cycleState *framework.CycleState, podToSchedule *v1.Pod, podToRemove *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
+	meta, err := getPodAffinityMetadata(cycleState)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	meta.UpdateWithPod(podToRemove, podToSchedule, nodeInfo.Node(), -1)
+	return nil
+}
+
+func getPodAffinityMetadata(cycleState *framework.CycleState) (*predicates.PodAffinityMetadata, error) {
+	c, err := cycleState.Read(preFilterStateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	s, ok := c.(*preFilterState)
+	if !ok {
+		return nil, fmt.Errorf("%+v  convert to interpodaffinity.state error", c)
+	}
+	return s.meta, nil
+}
+
 // Filter invoked at the filter extension point.
 func (pl *InterPodAffinity) Filter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
-	meta, ok := migration.CovertStateRefToPredMeta(migration.PredicateMetadata(cycleState))
-	if !ok {
-		return migration.ErrorToFrameworkStatus(fmt.Errorf("%+v convert to predicates.Metadata error", cycleState))
+	// We ignore the error for now since we are able to handle the nil metadata.
```

Comment:
Why ignore the error? We can add a step in `getPodAffinityMetadata()`, Line125, to return {nil, nil}, so that all error returned are serious errors.

```go
	if c == nil {
		return nil, nil
	}
```

Reply:
As I mentioned in  https://github.com/kubernetes/kubernetes/pull/86175#discussion_r356903674, CycleSate.Read returns an error when the entry is not found: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/v1alpha1/cycle_state.go#L95

we don't want to error on that. But I guess we can error on the failed cast, and so we return nil if Read returned an error, please check the latest patch.